### PR TITLE
Azure.Core 2.0: Update AzureKeyCredential, RetryPolicy, and StatusCodeClassifier

### PR DESCRIPTION
**Pull request by annelo-msft**
## Description

Many of the basic Azure.Core types now have analogous types in the System.ClientModel library.  In order to follow the [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself) engineering principle, we plan to update Azure.Core to reuse functionality from these types, using either inheritance or adapters in most cases.

This PR makes the following changes:
- Azure.Core `AzureKeyCredential` inherits from System.ClientModel `ApiKeyCredential`
- Azure.Core `RetryPolicy` adapts System.ClientModel `ClientRetryPolicy`
- Azure.Core `StatusCodeClassifier` uses new internal implementation copied from System.ClientModel `BitVector640`

## Design Motivation

- `AzureKeyCredential` inherits from `ApiKeyCredential` to reuse the implementation.
- `RetryPolicy` wraps `ClientRetryPolicy` because inheritance is not possible here.
- `StatusCodeClassifier` does not reuse anything other than a new implementation we have been meaning to move it to for a while.

## Context

This PR is a smaller chunk of https://github.com/Azure/azure-sdk-for-net/pull/41773 to make it easier to review.  In addition, it this PR targets the [feature/azure.core-2.0](https://github.com/Azure/azure-sdk-for-net/tree/[feature/azure.core-2.0](https://github.com/Azure/azure-sdk-for-net/tree/feature/azure.core-2.0)) feature branch because we cannot move the Azure.Core 2.0 integration into main until after System.ClientModel 1.1.0 is GA.

### CODE REVIEW

File: sdk/core/Azure.Core/src/Internal/BitVector640.cs [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+/// into ten ulongs, where each bit of the ulong represents a single bool value.
+///
+/// It exposes a public indexer so that the bool values can be accessed as a
+/// standard .NET collection API.
+///
+/// It is used in System.ClientModel and Azure.Core to implement response
+/// classifiers, where each true bit represents a status code that the
+/// classifier considers a success code.
+/// </summary>
+internal struct BitVector640
```
Comment (annelo-msft) [inline sdk/core/Azure.Core/src/Internal/BitVector640.cs:19]: Copied in from System.ClientModel/Internal.
Removing duplication across Azure.Core and System.ClientModel is tracked by https://github.com/Azure/azure-sdk-for-net/issues/41035

File: sdk/core/Azure.Core/src/AzureKeyCredential.cs [hunk 2]
```diff
@@ -1,28 +1,30 @@
 ﻿// Copyright (c) Microsoft Corporation. All rights reserved.
 // Licensed under the MIT License.
 
+using System.ClientModel;
```
Comment (annelo-msft) [inline sdk/core/Azure.Core/src/AzureKeyCredential.cs:4]: This one has a small enough implementation surface that if we didn't want to make it inherit, I don't think that would be a problem.

File: sdk/core/System.ClientModel/src/Message/PipelineMessage.cs [hunk 3]
```diff
@@ -146,7 +146,7 @@ public void Apply(RequestOptions options)
     /// <param name="key">The key for the property in the message's property
     /// bag.</param>
     /// <param name="value">The value of the property.</param>
-    public void SetProperty(Type key, object value) =>
+    public void SetProperty(Type key, object? value) =>
```
Comment (annelo-msft) [inline sdk/core/System.ClientModel/src/Message/PipelineMessage.cs:149]: If we want to be able to set properties on PipelineMessage to `null` we'll need to make this nullable.  The changes to ArrayBackedPropertyBag accommodate this change.

If we decide to keep this API change, I'll add it to https://github.com/Azure/azure-sdk-for-net/pull/42328.

File: sdk/textanalytics/Azure.AI.TextAnalytics/src/TextAnalyticsFailedDetailsParser.cs [hunk 4]
```diff
@@ -20,84 +20,73 @@ public override bool TryParse(Response response, out ResponseError? error, out I
             data = default;
             error = default;
 
-            if (response.ContentStream is { CanSeek: true })
```
Comment (annelo-msft) [inline sdk/textanalytics/Azure.AI.TextAnalytics/src/TextAnalyticsFailedDetailsParser.cs:23]: Takes advantage of new buffering logic introduced in https://github.com/Azure/azure-sdk-for-net/pull/42294

File: sdk/core/Azure.Core/src/AzureKeyCredential.cs (hunk 2 above)
Comment (jsquire) [reply sdk/core/Azure.Core/src/AzureKeyCredential.cs:4]: Personally, I'm not sure that I see much benefit to inheritance other than consistency with other patterns.   I don't have a strong opinion either way.

File: sdk/core/Azure.Core/src/Internal/BitVector640.cs [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+    private ulong _bits4;
+    private ulong _bits5;
+    private ulong _bits6;
+    private ulong _bits7;
+    private ulong _bits8;
+    private ulong _bits9;
+
+    public bool this[int i]
+    {
+        readonly get
```
Comment (jsquire) [inline sdk/core/Azure.Core/src/Internal/BitVector640.cs:35]: That's interesting.   I don't know that I've ever seen this modifier pattern before.

File: sdk/core/Azure.Core/src/StatusCodeClassifier.cs [hunk 6]
```diff
@@ -14,91 +12,56 @@ namespace Azure.Core
...
-            }
-            else
-            {
-                value &= ~mask;
-            }
-
-            _successCodes[index] = value;
-        }
-
-        private bool IsSuccessCode(int statusCode)
```
Comment (jsquire) [inline sdk/core/Azure.Core/src/StatusCodeClassifier.cs:94]: What kind of crazy magic is this?  I'm both fascinated and afraid of it.

File: sdk/core/System.ClientModel/src/Message/PipelineMessage.cs (hunk 3 above)
Comment (jsquire) [reply sdk/core/System.ClientModel/src/Message/PipelineMessage.cs:149]: What's the scenario where we'd want to set properties to null?   
_(I don't have an opinion, I'm just curious about the end-to-end)_

**Review by jsquire — APPROVED** [review]

File: sdk/core/Azure.Core/src/StatusCodeClassifier.cs (hunk 6 above)
Comment (annelo-msft) [reply sdk/core/Azure.Core/src/StatusCodeClassifier.cs:94]: If the comments here are insufficient to answer your question, let me know - I 'm happy to unpack it more as needed: https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/core/System.ClientModel/src/Internal/BitVector640.cs#L35-L62

File: sdk/core/Azure.Core/src/Pipeline/RetryPolicy.cs [hunk 7]
```diff
@@ -184,88 +100,185 @@ internal virtual void Wait(TimeSpan time, CancellationToken cancellationToken)
...
+            }
+
+            protected override void OnRequestSent(PipelineMessage message)
+            {
+                _azureCorePolicy.OnRequestSent(HttpMessage.GetHttpMessage(message));
+
+                if (!message.TryGetProperty(typeof(BeforeTimestamp), out object? beforeTimestamp) ||
+                    beforeTimestamp is not long before)
+                {
+                    throw new InvalidOperationException("'BeforeProcessTimestamp' was not set on message by RetryPolicy.");
```
Comment (christothes) [inline sdk/core/Azure.Core/src/Pipeline/RetryPolicy.cs:212]: Is this possible? Would it be better to log here than to throw - I think I'd rather have the retry execute without timing data than to hard fail.

Comment (annelo-msft) [reply sdk/core/Azure.Core/src/Pipeline/RetryPolicy.cs:212]: That's a good point.  I don't think it is, but I'll change it to a Debug.Assert and reset the values to something that makes more sense.  Thanks!

**Review by christothes — APPROVED** [review]

File: sdk/core/Azure.Core/src/AzureKeyCredential.cs (hunk 2 above)
Comment (KrzysztofCwalina) [reply sdk/core/Azure.Core/src/AzureKeyCredential.cs:4]: One advantage I see is that if we inherit, we will be able to use one policy to inject the key into the request.

File: sdk/core/Azure.Core/src/StatusCodeClassifier.cs (hunk 6 above)
Comment (KrzysztofCwalina) [reply sdk/core/Azure.Core/src/StatusCodeClassifier.cs:94]: > What kind of crazy magic is this? I'm both fascinated and afraid of it.

branchless programming: https://en.algorithmica.org/hpc/pipelining/branchless/

File: sdk/core/Azure.Core/src/AzureKeyCredential.cs (hunk 2 above)
Comment (annelo-msft) [reply sdk/core/Azure.Core/src/AzureKeyCredential.cs:4]: Good point.  We will inherit.

File: sdk/core/Azure.Core/src/StatusCodeClassifier.cs (hunk 6 above)
Comment (annelo-msft) [reply sdk/core/Azure.Core/src/StatusCodeClassifier.cs:94]: @KrzysztofCwalina, IIRC the switch statement makes it not 100% branchless - I read the IL a few months ago.  But I didn't spend the time to figure out how to do better for this case.

File: sdk/core/System.ClientModel/src/Message/PipelineMessage.cs (hunk 3 above)
Comment (annelo-msft) [reply sdk/core/System.ClientModel/src/Message/PipelineMessage.cs:149]: Answered elsewhere - to un-set properties without adding another member.

**Pull request merged**